jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/Unit.java

```diff
@@ -0,0 +1,249 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.deploy.providers;
+
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.eclipse.jetty.deploy.App;
+import org.eclipse.jetty.util.StringUtil;
+import org.eclipse.jetty.util.resource.PathCollators;
+
+/**
+ * A Unit of deployment, a basename and all the associated
+ * paths that belong to that basename, along with state.
+ */
+public class Unit
+{
+    public enum State
+    {
+        UNCHANGED,
+        ADDED,
+        CHANGED,
+        REMOVED
+    }
+
+    private final String baseName;
+    private final Map<Path, State> paths = new HashMap<>();
+    /**
+     * Indicates a synthetic Unit, used to track Environment specific configurations.
+     * Such as {@code <env>.xml}, {@code <env>.properties}, {@code <env>-<desc>.xml}, {@code <env>-<desc>.properties},
+     * all under a single unit that will not actually be deployed.
+     */
+    private String envConfig;
+    private App app;
+    private State state;
+
+    public Unit(String basename)
+    {
+        this.baseName = basename;
+        this.state = calcState();
+    }
+
+    public String getBaseName()
+    {
+        return baseName;
+    }
+
+    public State getState()
+    {
+        return state;
+    }
+
+    /**
+     * <p>
+     * Calculate the State of the overall Unit based on the States in the Paths.
+     * </p>
+     * <dl>
+     * <dt>UNCHANGED</dt>
+     * <dd>All Path states are in UNCHANGED state</dd>
+     * <dt>ADDED</dt>
+     * <dd>All Path states are in ADDED state</dd>
+     * <dt>CHANGED</dt>
+     * <dd>At least one Path state is CHANGED, or there is a variety of states</dd>
+     * <dt>REMOVED</dt>
+     * <dd>All Path states are in REMOVED state, or there are no Paths being tracked</dd>
```

Comment:
again,  if all deployable files (xml, war, directory) are removed, leaving only a property file, then the result should be REMOVED even though all paths are not removed.

I think this algorithm definitely needs the concept of deployable files (xml, war, directory)